src/service/layers-impl.js

```diff
@@ -0,0 +1,634 @@
+   * @param {!Element} element
+   * @return {!LayoutElement}
+   */
+  declareLayer_(element) {
+    const layout = this.add(element);
+    layout.declareLayer();
+    return layout;
+  }
+
+  onResize_() {
+    const layouts = this.layouts_;
+    for (let i = 0; i < layouts.length; i++) {
+      const layout = layouts[i];
+      layout.undeclareLayer();
+      layout.forgetParentLayer();
+    }
+  }
+
+  scrolled_(element) {
+    let layout = LayoutElement.forOptional(element);
+    if (layout && layout.isLayer()) {
+      layout.requestScrollRemeasure();
+    } else {
+      layout = this.declareLayer_(element);
+    }
+
+    if (this.onScroll_) {
+      this.onScroll_(/* layer.getElements() */);
+    }
+  }
+
+  onScroll(handler) {
+    this.onScroll_ = handler;
+  }
+}
+
+export class LayoutElement {
+  constructor(element) {
+    element[LAYOUT_PROP] = this;
+    /**
+     * @private @const {!Element}
+     */
+    this.element_ = element;
+
+    /** @type {?LayoutElement|undefined} */
+    this.parentLayer_ = undefined;
+
+    this.needsRemeasure_ = true;
+    this.size_ = SizeWh(0, 0);
+    this.position_ = PositionLt(0, 0);
+
+    this.isLayer_ = false;
+    this.isRootLayer_ = false;
+    this.needsScrollRemeasure_ = false;
+    this.scrollLeft_ = 0;
+    this.scrollTop_ = 0;
+
+    this.children_ = [];
+  }
+
+  /**
+   * Gets the element's LayoutElement instance.
+   * @param {!Element}
+   * @return {!LayoutElement}
+   */
+  static for(element) {
+    return dev().assert(LayoutElement.forOptional(element));
+  }
+
+  /**
+   * Gets the element's LayoutElement instance, if the element has one.
+   * @param {!Element}
+   * @return {?LayoutElement}
+   */
+  static forOptional(element) {
+    return element[LAYOUT_PROP];
+  }
+
+  /**
+   * Finds the element's parent layer
+   * If the element is itself a layer, it returns the layer's parent layer.
+   * @param {!Element}
+   * @param {opt_force=} Whether to force a re-lookup
+   * @return {?LayoutElement}
+   */
+  static getParentLayer(element, opt_force) {
+    if (!opt_force) {
+      const layout = LayoutElement.forOptional(element);
+      if (layout) {
+        return layout.getParentLayer();
+      }
+    }
+
+    const win = element.ownerDocument.defaultView;
+    if (computedStyle(win, element).position == 'fixed') {
+      LayoutLayers.declareLayer(element);
+      return null;
+    }
+
+    let last = element;
+    let op = element.offsetParent;
+    for (let el = last.parentNode; el; el = el.parentNode) {
+      const layout = LayoutElement.forOptional(el);
+      if (layout && layout.isLayer()) {
+        return layout;
+      }
+
+      // now check to see if offsetParent is a fixed layer
+      if (el === op) {
+        if (computedStyle(win, op).position == 'fixed') {
+          LayoutLayers.declareLayer(op);
+          return op;
+        }
+        op = op.offsetParent;
+      }
+
+      last = el;
+    }
+
+
+    if (last.nodeType !== Node.DOCUMENT_NODE) {
```

Comment:
This could also be a shadow root. I think we don't need this check. If you want to assert that the element is in DOM, you can call `node.isConnected` (https://github.com/ampproject/amphtml/blob/master/src/service/ampdoc-impl.js#L149). It's not available on all browsers, but available on Chrome and good enough for a dev error. Otherwise, certainly checking for `document || shadowRoot` is probably good enough.

Another nuance: what if this is a FIE? Do you traverse over the boundary of FIE? E.g. `parent = element.ownerDocument.defaultView.frameElement`? E.g. see https://github.com/ampproject/amphtml/blob/master/src/service/ampdoc-impl.js#L183

Reply:
I guess it's more of a system-wide question: how would layers treat FIE. E.g. we could automatically assign a layer to the FIE's `document` or `docomentElement` and then calculate parent via `frameElement`. Then the FIE will be in the overall layer hierarchy. Or any other approach possible. But there should be some provisioning for FIE. Otherwise, each FIE will be its own root layer and we can loose the overall position, me thinks.

Reply:
Agreed. Is there a common way to go from embed (either FIE or shadow) to the host document where they're embedded? Ideally, I'll check if the topmost node inside is an embed, and call `getParentLayer` again with the embed's host element.

Reply:
Once you reach the root of the FIE (e.g. `documentElement` or `document` node) you should be able to get the frame element to go across the boundary in the parent doc. See https://github.com/ampproject/amphtml/blob/master/src/service.js#L337